### ISSUE TITLE
Segmentation Fault on Ubuntu 20.04

### ISSUE THREAD

**Issue by maitiSoutrik** [OneLoneCoder/olcPixelGameEngine#237]
Hello,
I am trying to compile the Hello World script using 
g++ -o play HelloWorld.cpp -lX11 -lGL -lpthread -lpng -lstdc++fs -std=c++17
 of the pixel game engine on Ubuntu 20.04 and I am getting the following error when I try to run the executable using ./nameOfExecutable 
 
```
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x0000555555569fe8 in olc::Platform_Linux::CreateWindowPane(olc::v2d_generic<int> const&, olc::v2d_generic<int>&, bool) ()
```
Am I missing something?
I would really appreciate some help.

**Comment by Moros1138** [comment]
SIGSEGV means that your program has experienced a segmentation fault. In general, this happens when you try to access memory improperly.

An example might be, trying at access memory that's been freed, dereferencing a null pointer, accessing memory that is out of bounds, or writing into read-only memory.

It's hard to give more targeted advice without seeing some code. and having some more information about your build environment.

**Comment by maitiSoutrik** [comment]
Thank you Moros1138 for the quick reply.
The following is the code I am trying to compile :
```
#define OLC_PGE_APPLICATION
#include "olcPixelGameEngine.h"
#include <iostream>

class Example : public olc::PixelGameEngine {
public:
    Example(){
        sAppName = "Example";
    }
    
    bool OnUserCreate() override {
        // called once at the start, so create things here
        return true;
    }

    bool OnUserUpdate(float fElapsedTime) override {
        // called once per frame
        for (int x = 0; x < ScreenWidth(); x++)
            for (int y = 0; y < ScreenHeight(); y++)
                Draw(x, y, olc::Pixel(rand() % 256, rand() % 256, rand() % 256));
        return true;
    }

};

int main()
{
    Example demo;
    if (demo.Construct(160, 160, 4, 4))
        demo.Start();

    return 0;
    //std::cout << "Hello World!\n";
}
```
I would really appreciate if you could tell me what exact build information you are looking for.
Does the above code snippet work ?
I also wanted to point out that the olcpixelgameengine.h and the cpp file are in the same folder.

**Comment by Moros1138** [comment]
ok.. time for knitty gritty stuff

can you please post the output from this command

```
g++ -v
```

also can you confirm which version of PGE you're using, it should be prominently visible near the top of the header file.

I'm grasping at straws here, to be honest, if that's all your code, i see no reason for it to segfault.

**Comment by maitiSoutrik** [comment]
Hi @Moros1138 ,
This is the output of g++ -v :
```
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/9/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 9.3.0-17ubuntu1~20.04' --with-bugurl=file:///usr/share/doc/gcc-9/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,gm2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-9 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-9-HskZEa/gcc-9-9.3.0/debian/tmp-nvptx/usr,hsa --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04) 
```

The PGE version is as follows : 
```
#pragma region license_and_help
/*
	olcPixelGameEngine.h

	+-------------------------------------------------------------+
	|           OneLoneCoder Pixel Game Engine v2.15              |
	|  "What do you need? Pixels... Lots of Pixels..." - javidx9  |
	+-------------------------------------------------------------+
```

**Comment by Moros1138** [comment]
Can you describe your machine, it's components, your graphics card, the resolution of your monitor.. I'm really at the point where i'm looking around at anything that could be an issue.

**Comment by maitiSoutrik** [comment]
Hello @Moros1138 ,
Thanks for such a quick reply.
The machine on which I am trying to compile this is a server host.
I have an Intel NUC with a i7 processor and I have installed Ubuntu server on it.
I usually ssh -X into it since it doesn't have any displays. I use gui based programs easily using this command but I don't know about PGE. Maybe I need to change some setting?

**Comment by Moros1138** [comment]
No display? .. ok.. then it's possible you don't have graphics drivers installed. PGE requires at the very least opengl 1.0, which requires a GPU and it's drivers to be installed. If drivers are not installed, or the GPU doesn't support opengl that would cause you the PGE application to have a segfault, as PGE makes the assumption that you're not running on (and i mean this in good humor) a potato environment.

if you have a gpu, try figuring out what it is and get some drivers installed for it.

**Comment by maitiSoutrik** [comment]
Hi @Moros1138 ,
I tried doing a ssh -Y and now I am getting a fresh set of different errors :
```
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  149 (GLX)
  Minor opcode of failed request:  6 (X_GLXIsDirect)
  Serial number of failed request:  38
  Current serial number in output stream:  37
```
I am trying to access this host on my Mac which already has XQuartz installed and updated for viewing applications with gui over ssh.
I also tried installing the libGL libraries using the following :
```
sudo apt-get install -y mesa-utils
sudo apt-get install -y libgl1-mesa-glx
```
But I am still getting the above error.

**Comment by Moros1138** [comment]
I'm completely out of where I'm confident giving advice, so from here on out any advice I do give is speculative and should be taken with caution.

check the output of the following command:
``xrandr``

it should give you a list of your display capabilities, example output looks like this:

![](https://i.imgur.com/Ls5DaCZ.png)

if you don't have it, it's easy enough to install
``sudo apt-get install libxrandr2``

**Comment by maitiSoutrik** [comment]
Hi @Moros1138 ,
I really appreciate your help but this [article](https://unix.stackexchange.com/questions/429760/opengl-rendering-with-x11-forwarding) me fixed the issues I was having.
Now it works properly.

**Comment by Moros1138** [comment]
Glad you're all sorted out.. So this was a problem client side, not with the X server on Ubuntu?

**Comment by maitiSoutrik** [comment]
I had to do a couple of things :
1. I ran the following on my mac terminal
`defaults write org.xquartz.X11 enable_iglx -bool true`
2. Then I tested if I am able to run any graphical application from the server using ssh -Y
3. Then I run this **glxgears** to see the moving gears application
4. Then I run the executable that I made using PGE and it works like it is supposed to

So Yeah I guess the problem was on the client side
